src/stdlib/RandomMeasures.jl

```diff
@@ -128,41 +128,31 @@ function distribution(d::SizeBiasedSamplingProcess{<:DirichletProcess})
     return LocationScale(zero(α), d.surplus, Beta(one(α), α))
 end
 
-function _logpdf_table(d::DirichletProcess, m::AbstractVector{Int})
+function _logpdf_table(d::DirichletProcess{T}, m::AbstractVector{Int}) where {T<:AbstractFloat}
     # compute the sum of all cluster counts
     sum_m = sum(m)
 
-    # shortcut if all cluster counts are zero
-    dα = d.α
-    T = typeof(dα)
-    iszero(sum_m) && return zeros(T, 1)
-
     # pre-calculations
-    z = log(sum_m - 1 + dα)
+    z = log(sum_m - 1 + d.α)
 
     # construct the table
-    K = length(m)
-    table = Vector{T}(undef, K)
-    contains_zero = false
-    @inbounds for i in 1:K
-        mi = m[i]
-
-        if iszero(mi)
-            if contains_zero
-                table[i] = -Inf
-            else
-                table[i] = log(dα) - z
-                contains_zero = true
-            end
-        else
-            table[i] = log(mi) - z
-        end
+    has_zeros = any(i -> i == 0, m)
+    K = has_zeros ? length(m) : length(m)+1
+    table = ones(T, K) * T(-Inf)
+
+    if iszero(sum_m)
+        table[1] = T(0)
+        return table
     end
 
-    if !contains_zero
-        push!(table, log(dα) - z)
+    contains_zero = false
+    @inbounds for i in (has_zeros ? findall(j -> j>0, m) : 1:(K-1))
+        table[i] = T(log(m[i]) - z)
     end
 
+    k_new = has_zeros ? findfirst(j -> j == 0, m) : K
```

Comment:
Again I think one could use `iszero` here. Moreover, I would suggest computing `first_zero = findfirst(iszero, m)` at the beginning of the function, since that yields `has_zeros` naturally, without checking `m` twice. I.e., I would recommend to change the implementation above to
```julia
first_zero = findfirst(iszero, m)
has_zeros = first_zero !== nothing
```
and just defining
```julia
k_new = first_zero === nothing ? K : first_zero
```
here.

Reply:
Actually, I think `has_zeros` is not needed anymore, if you follow this suggestion and change also the `for` loop above since `K` can be defined based on `first_zero` as
```julia
K = first_zero === nothing ? length(m) + 1 : length(m)
```